File: backend/entities/books.py
import psycopg2
from ..connection.config import connect_db
from flask import request, jsonify

def get_books():
    try:
        conn = connect_db()
        if conn:
            cur = conn.cursor()

            # Consulta para buscar os livros e verificar a disponibilidade
            cur.execute("""
                SELECT b.book_id, b.title, b.author_id, b.publication_year, b.genre,
                CASE
                    WHEN a.book_id IS NOT NULL AND a.return_date IS NULL THEN 'Alugado'
                    ELSE 'Disponível'
                END AS disponibilidade
                FROM bibliotecas b
                LEFT JOIN alugueis a ON b.book_id = a.book_id AND a.return_date IS NULL;
            """)

            books = cur.fetchall()
            cur.close()
            conn.close()

            books_list = []
            for book in books:
                books_list.append({
                    'book_id': book[0],
                    'title': book[1],
                    'author_id': book[2],
                    'publication_year': book[3],
                    'genre': book[4],
                    'disponibilidade': book[5]  # Adiciona a coluna de disponibilidade ("Alugado" ou "Disponível")
                })

            return jsonify({'books': books_list}), 200

        else:
            print("Erro ao conectar ao banco de dados.")
            return jsonify({'error': 'Erro ao conectar ao banco de dados'}), 500

    except psycopg2.Error as e:
        print(f"Erro ao listar livros: {e}")
        return jsonify({'error': f"Erro ao listar livros: {str(e)}"}), 500
    except Exception as e:
        print(f"Erro geral: {e}")
        return jsonify({'error': f"Erro geral: {str(e)}"}), 500
    
def addBook(title, author_id, publication_year, genre):
    conn = connect_db()  # Conectar ao banco de dados
    if conn:
        try:
            cur = conn.cursor()
            cur.execute("""
                INSERT INTO Bibliotecas (title, author_id, publication_year, genre)
                VALUES (%s, %s, %s, %s);
            """, (title, author_id, publication_year, genre))
            conn.commit()  # Confirma a transação
            cur.close()  # Fechar o cursor
            conn.close()  # Fechar a conexão
            return jsonify({'data': 'Livro adicionado com sucesso!', 'status': 201})
        except psycopg2.Error as e:
            print(f"Erro ao adicionar Livro: {e}")
            return jsonify({'data': f"Erro ao adicionar Livro: {str(e)}", 'status': 500})
    else:
        return jsonify({'data': 'Erro ao conectar ao banco de dados', 'status': 500})



def deleteBook(book_id):
    conn = connect_db()
    if conn:
        try:
            cur = conn.cursor()
            cur.execute("DELETE FROM Bibliotecas WHERE book_id = %s;", (book_id,))
            conn.commit()
            print("Livro deletado com sucesso!")
        except psycopg2.Error as e:
            print(f"Erro ao deletar o livro: {e}")
        finally:
            cur.close()
            conn.close()
    else:
        print("Não foi possível conectar ao banco de dados.")


def update_book_in_db(book_id, title=None, author_id=None, publication_year=None, genre=None):
    conn = connect_db()
    if conn:
        try:
            cur = conn.cursor()

            # Adicionando logs para verificar se os dados estão corretos
            print(f"Atualizando livro com ID {book_id}:")
            print(f"Título: {title}, Autor: {author_id}, Ano: {publication_year}, Gênero: {genre}")

            # Query de atualização no PostgreSQL
            cur.execute("""
                UPDATE Bibliotecas 
                SET title = %s, author_id = %s, publication_year = %s, genre = %s
                WHERE book_id = %s;
            """, (title, author_id, publication_year, genre, book_id))

            conn.commit()  # Verifique se o commit está sendo chamado
            print(f"Livro com ID {book_id} atualizado com sucesso.")
            return True
        except psycopg2.Error as e:
            print(f"Erro ao atualizar o livro: {e}")
            return False
        finally:
            cur.close()
            conn.close()
    else:
        print("Erro ao conectar ao banco de dados.")
        return False



def search_book(search_term, search_type):
    conn = connect_db()
    if conn:
        try:
            cur = conn.cursor()
            if search_type == "title" or search_type == "genre":
                query = f"SELECT * FROM Bibliotecas WHERE {search_type} ILIKE %s;"
                cur.execute(query, (f"%{search_term}%",))
            elif search_type == "author_id" or search_type == "publication_year":
                query = f"SELECT * FROM Bibliotecas WHERE {search_type} = %s;"
                cur.execute(query, (search_term,))
            
            books = cur.fetchall()

            if books:
                for book in books:
                    print(f"ID: {book[0]}, Título: {book[1]}, ID do Autor: {book[2]}, Ano de Publicação: {book[3]}, Gênero: {book[4]}")
            else:
                print("Nenhum livro encontrado.")
        except psycopg2.Error as e:
            print(f"Erro ao consultar o livro: {e}")
        finally:
            cur.close()
            conn.close()
    else:
        print("Não foi possível conectar ao banco de dados.")


def rent_book(member_id, book_id, rental_date, return_date):
    try:
        conn = connect_db()
        if conn:
            cur = conn.cursor()

            print(f"Iniciando o aluguel para o livro {book_id} e membro {member_id}")

            # Verificar se o livro já está alugado usando a função auxiliar
            if is_book_rented(book_id):
                print(f"Livro {book_id} já está alugado.")
                return {'data': 'Este livro já está alugado e não pode ser alugado novamente até ser devolvido.', 'status': 400}

            # Inserir o aluguel na tabela alugueis
            cur.execute("""
                INSERT INTO alugueis (member_id, book_id, rental_date, return_date)
                VALUES (%s, %s, %s, %s);
            """, (member_id, book_id, rental_date, return_date))

            # Atualizar o status do livro para 'Alugado'
            cur.execute("""
                UPDATE bibliotecas SET disponibilidade = 'Alugado' WHERE book_id = %s;
            """, (book_id,))

            conn.commit()  # Confirma a transação
            print(f"Aluguel do livro {book_id} registrado com sucesso para o membro {member_id}!")

            cur.close()  # Fecha o cursor
            conn.close()  # Fecha a conexão

            return {'data': 'Livro alugado com sucesso!', 'status': 200}

        else:
            print("Erro ao conectar ao banco de dados")
            return {'data': 'Erro ao conectar ao banco de dados', 'status': 500}

    except psycopg2.Error as e:
        print(f"Erro ao alugar livro no banco de dados: {str(e)}")
        return {'data': f"Erro ao alugar livro: {str(e)}", 'status': 500}

def return_book(book_id):
    try:
        conn = connect_db()
        if conn:
            cur = conn.cursor()

            # Verificar se o livro está alugado (se existe um registro de aluguel sem data de devolução)
            cur.execute("SELECT rental_id FROM alugueis WHERE book_id = %s AND return_date IS NULL;", (book_id,))
            rental = cur.fetchone()

            if rental:
                # Atualizar a data de devolução
                cur.execute("UPDATE alugueis SET return_date = NOW() WHERE rental_id = %s;", (rental[0],))
                conn.commit()

                # Atualizar a disponibilidade do livro para 'Disponível'
                cur.execute("UPDATE bibliotecas SET disponibilidade = 'Disponível' WHERE book_id = %s;", (book_id,))
                conn.commit()

                cur.close()
                conn.close()

                print(f"Livro com ID {book_id} devolvido com sucesso!")
                return {'data': 'Livro devolvido com sucesso!', 'status': 200}
            else:
                print(f"Livro com ID {book_id} já foi devolvido ou não estava alugado.")
                return {'data': 'Livro já devolvido ou não está alugado.', 'status': 400}

    except psycopg2.Error as e:
        print(f"Erro ao devolver o livro no banco de dados: {str(e)}")
        return {'data': f"Erro ao devolver o livro: {str(e)}", 'status': 500}

    
def is_book_rented(book_id):
    conn = connect_db()
    if conn:
        try:
            cur = conn.cursor()
            
            # Verifica se o livro está alugado (return_date = NULL)
            cur.execute("SELECT rental_id FROM alugueis WHERE book_id = %s AND return_date IS NULL;", (book_id,))
            rental = cur.fetchone()
            
            cur.close()
            conn.close()

            # Retorna True se o livro estiver alugado
            return rental is not None
        except psycopg2.Error as e:
            print(f"Erro ao verificar se o livro está alugado: {e}")
            return False
    else:
        print("Erro ao conectar ao banco de dados.")
        return False

File: frontend/js/books.js
document.getElementById('refreshBooks').addEventListener('click', clearAndReloadBooks);
document.getElementById('cancelDelete').addEventListener('click', hidePopover);
document.getElementById('rentBookButton').addEventListener('click', showRentPopover)
//Função para apagar conteúdo da tabela e recarregar a página
function clearAndReloadBooks() {
    const tableBody = document.getElementById("booksTableBody");
    tableBody.innerHTML = ""; // Apaga o conteúdo da tabela
    getBooks(); // Recarrega os dados
}

// Função para abrir o modal de edição e preencher os campos com os dados do livro
function openEditModal(bookId, title, authorId, publicationYear, genre) {
    document.getElementById('editBookId').value = bookId;
    document.getElementById('editTitle').value = title;
    document.getElementById('editAuthorId').value = authorId;
    document.getElementById('editPublicationYear').value = publicationYear;
    document.getElementById('editGenre').value = genre;
    document.getElementById('editBookModal').style.display = 'block';
}

function openDeletePopover(bookId) {
    const popoverDelete = document.getElementById('deleteBookPopover');
    const popoverDeleteBtn = document.getElementById('popoverDeleteBtn');
    const popoverDeleteMsg = document.getElementById('popoverDeleteMsg');
    
    popoverDeleteMsg.innerHTML = 'Deseja deletar o livro com ID: ' + bookId + "?";

    popoverDeleteBtn.addEventListener('click', () => {
        deleteBook(bookId); // Chama a função de deletar o livro
        popoverDelete.style.display = 'none'; // Esconde o popover após deletar
    });

    popoverDelete.style.display = 'block'; // Mostra o popover ao clicar em deletar
}

// Função para fechar o modal de edição
function closeBookModal() {
    document.getElementById('editBookModal').style.display = 'none';
    hidePopover();
}

// Função para fechar o popover de deletar livro
function hidePopover() {
    const popover = document.getElementById('deleteBookPopover');
    if (popover) {
        popover.style.display = 'none';  // Esconde o popover de exclusão
    } else {
        console.error('Popover element not found');
    }
}

// Função para abrir o popover de aluguel de livro
function showRentPopover() {
    const rentPopover = document.getElementById('rentBookPopover');
    rentPopover.style.display = 'block';
}

// Função para fechar o popover de aluguel de livro
function hideRentPopover() {
    const rentPopover = document.getElementById('rentBookPopover');
    rentPopover.style.display = 'none';
}

// Adiciona evento ao botão "Alugar Livro" para mostrar o popover
document.getElementById('rentBookButton').addEventListener('click', showRentPopover);



// Função para salvar atualizações no livro (PUT)
async function saveUpdatedBook() {
    const bookId = document.getElementById('editBookId').value;
    const title = document.getElementById('editTitle').value;
    const authorId = document.getElementById('editAuthorId').value;
    const publicationYear = document.getElementById('editPublicationYear').value;
    const genre = document.getElementById('editGenre').value;

    try {
        const response = await fetch(`http://localhost:5000/books`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                book_id: bookId,
                title: title,
                author_id: authorId,
                publication_year: publicationYear,
                genre: genre
            })
        });

        // Verifique a resposta do backend
        const data = await response.json();
        console.log("Resposta do servidor após atualização:", data);

        if (response.ok) {
            console.log("Livro atualizado com sucesso, recarregando tabela...");
            clearAndReloadBooks(); // Recarrega a lista de livros após a edição
            closeBookModal(); // Fecha o modal de edição após salvar
        } else {
            console.error('Erro ao atualizar livro. Status:', response.status);
        }
    } catch (error) {
        console.error('Erro ao atualizar livro:', error);
    }
}


// Função para deletar livros (DELETE)
async function deleteBook(bookId) {
    try {
        const response = await fetch('http://localhost:5000/books', {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ book_id: bookId })
        });

        if (response.ok) {
            alert('Livro deletado com sucesso');
            clearAndReloadBooks(); // Recarrega a lista de livros após deletar
        } else {
            console.error('Erro ao deletar livro.');
        }
    } catch (error) {
        console.error('Erro ao deletar livro:', error);
    }
}


async function confirmRentBook() {
    const memberId = document.getElementById('rentMemberId').value;
    const bookId = document.getElementById('rentBookId').value;
    const rentalDate = document.getElementById('rentalDate').value;
    const returnDate = document.getElementById('returnDate').value || null;  // Retorna null se o campo estiver vazio

    if (!memberId || !bookId || !rentalDate) {
        alert("Todos os campos são obrigatórios.");
        return;
    }

    try {
        const response = await fetch('http://localhost:5000/alugueis', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                member_id: memberId,
                book_id: bookId,
                rental_date: rentalDate,
                return_date: returnDate
            })
        });

        const res = await response.json();

        if (response.ok) {
            alert(res.data || "Livro alugado com sucesso!");

            // Atualizar a célula de disponibilidade para "Alugado" apenas para o livro alugado
            const rows = document.querySelectorAll("#booksTableBody tr");
            rows.forEach(row => {
                const idCell = row.querySelector("td:first-child").textContent;
                if (idCell == bookId) {
                    const availabilityCell = row.querySelector("td:last-child");
                    availabilityCell.textContent = "Alugado";  // Atualiza o status para "Alugado"
                }
            });

            hideRentPopover();  // Esconde o popover após o sucesso
        } else {
            alert(res.data || "Erro ao alugar o livro.");
        }
    } catch (error) {
        console.error('Erro ao alugar o livro:', error);
        alert("Erro de conexão ou erro inesperado.");
    }
}

// Função para atualizar a disponibilidade com base no rentalId e a returnDate
function updateBookAvailability(rentalId, bookId, returnDate) {
    const rows = document.querySelectorAll("#booksTableBody tr");
    rows.forEach(row => {
        const idCell = row.querySelector("td:first-child").textContent;

        if (idCell == bookId) {
            const availabilityCell = row.querySelector("td:last-child");

            if (returnDate === null || returnDate === "") {
                console.log(`Livro com ID ${bookId} encontrado. Atualizando status para 'Alugado'.`);
                availabilityCell.textContent = "Alugado";  // Sem data de devolução, o livro está alugado
            } else {
                console.log(`Livro com ID ${bookId} encontrado. Atualizando status para 'Disponível'.`);
                availabilityCell.textContent = "Disponível";  // Com data de devolução, o livro está disponível
            }
        }
    });
}

// Função para verificar e atualizar a disponibilidade do livro automaticamente
async function checkBookAvailability(bookId) {
    try {
        const response = await fetch(`http://localhost:5000/check_availability/${bookId}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            }
        });

        const res = await response.json();

        if (response.ok) {
            const availability = res.availability;

            // Atualiza o status na tabela do frontend
            const rows = document.querySelectorAll("#booksTableBody tr");
            rows.forEach(row => {
                const idCell = row.querySelector("td:first-child").textContent;
                if (idCell == bookId) {
                    const availabilityCell = row.querySelector("td:last-child");
                    availabilityCell.textContent = availability;
                }
            });
        } else {
            console.error(res.error || "Erro ao verificar disponibilidade.");
        }
    } catch (error) {
        console.error('Erro ao verificar disponibilidade:', error);
    }
}

// Chamada para verificar a disponibilidade de todos os livros ao carregar a página
document.addEventListener('DOMContentLoaded', () => {
    const bookIds = Array.from(document.querySelectorAll("#booksTableBody tr td:first-child")).map(td => td.textContent);
    bookIds.forEach(bookId => checkBookAvailability(bookId));  // Verifica todos os livros na tabela
});


// Função que aluga o livro e atualiza o status de disponibilidade
async function confirmRentBook() {
    const memberId = document.getElementById('rentMemberId').value;
    const bookId = document.getElementById('rentBookId').value;
    const rentalDate = document.getElementById('rentalDate').value;
    const returnDate = document.getElementById('returnDate').value || null;  // Retorna null se o campo estiver vazio

    if (!memberId || !bookId || !rentalDate) {
        alert("Todos os campos são obrigatórios.");
        return;
    }

    try {
        const response = await fetch('http://localhost:5000/alugueis', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                member_id: memberId,
                book_id: bookId,
                rental_date: rentalDate,
                return_date: returnDate
            })
        });

        const res = await response.json();

        if (response.ok) {
            alert(res.data || "Livro alugado com sucesso!");

            const rentalId = res.rental_id;
            const updatedAvailability = res.availability;

            // Atualizar o status de disponibilidade na tabela usando o rentalId e returnDate
            updateBookAvailability(rentalId, bookId, returnDate);

            hideRentPopover();  // Esconde o popover após o sucesso
        } else {
            alert(res.data || "Erro ao alugar o livro.");
        }
    } catch (error) {
        console.error('Erro ao alugar o livro:', error);
        alert("Erro de conexão ou erro inesperado.");
    }
}
async function fetchBooks() {
    try {
        const response = await fetch('http://localhost:5000/livros'); // Endpoint que retorna a lista de livros
        const books = await response.json();

        const tableBody = document.getElementById('booksTableBody');
        tableBody.innerHTML = ''; // Limpa a tabela antes de preencher

        books.data.forEach(book => {
            const row = document.createElement('tr');
            
            row.innerHTML = `
                <td>${book.id}</td>
                <td>${book.title}</td>
                <td>${book.author_id}</td>
                <td>${book.publication_year}</td>
                <td>${book.genre}</td>
                <td>
                    <button onclick="rentBook(${book.id})">Alugar</button>
                    <button onclick="editBook(${book.id})">Editar</button>
                    <button onclick="deleteBook(${book.id})">Excluir</button>
                </td>
                <td>${book.availability}</td>
            `;
            tableBody.appendChild(row);
        });

    } catch (error) {
        console.error('Erro ao buscar livros:', error);
        alert('Erro ao buscar livros.');
    }
}

// Chama fetchBooks ao carregar a página
window.onload = fetchBooks;


async function loadBooks() {
    try {
        const response = await fetch('http://localhost:5000/books', {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            }
        });

        const data = await response.json();
        if (response.ok) {
            const books = data.books;  // Acessa a chave 'books'

            const tableBody = document.getElementById('booksTableBody');
            tableBody.innerHTML = ''; // Limpa a tabela antes de carregar os novos dados

            books.forEach(book => {
                const row = document.createElement('tr');

                const bookIdCell = document.createElement('td');
                bookIdCell.textContent = book.book_id;

                const titleCell = document.createElement('td');
                titleCell.textContent = book.title;

                const statusCell = document.createElement('td');
                statusCell.textContent = book.disponibilidade;

                row.appendChild(bookIdCell);
                row.appendChild(titleCell);
                row.appendChild(statusCell);

                tableBody.appendChild(row);
            });
        } else {
            console.error('Erro ao carregar os livros:', data);
        }
    } catch (error) {
        console.error('Erro ao carregar os livros:', error);
    }
}



// Chame esta função quando a página carregar
document.addEventListener('DOMContentLoaded', loadBooks);


async function returnBook(rentalId) {
    try {
        const response = await fetch(`http://localhost:5000/books`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            }
        });

        if (response.ok) {
            const res = await response.json();
            alert(`Livro devolvido com sucesso! Multa: ${res.fine}`);
            getBooks(); // Recarregar os livros para atualizar a disponibilidade
        } else {
            console.error('Erro ao devolver o livro.');
        }
    } catch (error) {
        console.error('Erro ao devolver o livro:', error);
    }
}

// Função para adicionar livros (POST)
async function addBook() {
    const title = document.getElementById('title').value;
    const authorId = document.getElementById('author_id').value;
    const publicationYear = document.getElementById('publication_year').value;
    const genre = document.getElementById('genre').value;

    // Desabilita o botão para evitar múltiplos envios
    const saveButton = document.getElementById('save');
    saveButton.disabled = true;

    try {
        const response = await fetch('http://localhost:5000/books', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                title: title,
                author_id: authorId,
                publication_year: publicationYear,
                genre: genre
            })
        });

        if (response.ok) {
            clearAndReloadBooks(); // Recarrega a lista de livros após adicionar
            hidePopover(); // Fecha o popover
        } else {
            console.error('Erro ao adicionar livro.');
        }
    } catch (error) {
        console.error('Erro ao adicionar livro:', error);
    }
}


// Função para buscar livros no banco de dados (GET)
async function getBooks() {
    const url = "http://localhost:5000/books";
    try {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Erro ao buscar livros. Status: ${response.status}`);
        }
        const json = await response.json();
        const books = json.books;  // Acessa a chave 'books' diretamente

        console.log("Livros recebidos:", books);  // Verifique os dados recebidos

        const tableBody = document.getElementById("booksTableBody");

        // Limpa o conteúdo anterior
        tableBody.innerHTML = "";

        // Adiciona os novos livros na tabela
        books.forEach(book => {
            const row = document.createElement("tr");

            const idCell = document.createElement("td");
            idCell.textContent = book.book_id;

            const titleCell = document.createElement("td");
            titleCell.textContent = book.title;

            const authorIdCell = document.createElement("td");
            authorIdCell.textContent = book.author_id;

            const yearCell = document.createElement("td");
            yearCell.textContent = book.publication_year;

            const genreCell = document.createElement("td");
            genreCell.textContent = book.genre;

            const actionCell = document.createElement("td");

            const availabilityCell = document.createElement("td");
            availabilityCell.textContent = book.disponibilidade;

            // Botão para editar
            const editButton = document.createElement("button");
            editButton.textContent = "Editar";
            editButton.addEventListener('click', () => {
                openEditModal(book.book_id, book.title, book.author_id, book.publication_year, book.genre);
            });

            // Botão para deletar
            const deleteButton = document.createElement("button");
            deleteButton.textContent = "Deletar";
            deleteButton.addEventListener('click', () => {
                openDeletePopover(book.book_id);
            });

            actionCell.appendChild(deleteButton);
            actionCell.appendChild(editButton);
            row.appendChild(idCell);
            row.appendChild(titleCell);
            row.appendChild(authorIdCell);
            row.appendChild(yearCell);
            row.appendChild(genreCell);
            row.appendChild(actionCell);
            row.appendChild(availabilityCell);

            tableBody.appendChild(row);
        });

    } catch (error) {
        console.error('Erro ao buscar livros:', error);
    }
}

// Carregar os livros ao carregar a página
window.onload = function () {
    getBooks();

    // Adiciona o evento de busca na barra de pesquisa
    const searchInput = document.querySelector('#search input[name="q"]');
    searchInput.addEventListener('input', filterBooks);
};

// Função para filtrar livros com base no input de pesquisa
function filterBooks() {
    const searchInput = document.querySelector('#search input[name="q"]').value.toLowerCase();
    const booksTableBody = document.getElementById('booksTableBody');
    const rows = booksTableBody.getElementsByTagName('tr');

    for (let i = 0; i < rows.length; i++) {
        const titleCell = rows[i].getElementsByTagName('td')[1]; // Segunda coluna (Título)
        const genreCell = rows[i].getElementsByTagName('td')[4]; // Quinta coluna (Gênero)
        const title = titleCell.textContent.toLowerCase();
        const genre = genreCell.textContent.toLowerCase();

        // Verifica se o título ou o gênero contém o termo de busca
        if (title.includes(searchInput) || genre.includes(searchInput)) {
            rows[i].style.display = ''; // Mostra a linha
        } else {
            rows[i].style.display = 'none'; // Esconde a linha
        }
    }
}
